AndroidTest/app/src/main/java/cloudant/com/androidtest/MyActivity.java

```diff
@@ -49,77 +50,113 @@
+            mAdapter = new ArrayAdapter<TestResults>(this, R.layout.list_view_text);
 
-        Thread thread = new Thread(new Runnable() {
-            @Override
-            public void run() {
-                CloudantListener listener = new CloudantListener();
+            if (savedInstanceState != null || trs.size() > 0) {
+                testsRan = true;
+                mAdapter.addAll(trs.getAll());
+            }
 
-                JUnitCore core = new JUnitCore();
-                core.addListener(listener);
+            this.setListAdapter(mAdapter);
+
+            Thread thread = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    CloudantListener listener = new CloudantListener();
+
+                    JUnitCore core = new JUnitCore();
+                    core.addListener(listener);
+
+                    ArrayList<Class> classes = new ArrayList<Class>();
+                    for (Class c : BuildConfig.classToTests) {
+                        //Check parameter for specifying which test case package to run
+                        String testsToRun = String.valueOf(System.getProperty("test.run-packages"));
+                        //Run all classes if JVM parameter for specifying packages does not exist
+                        if (!testsToRun.equals("null")) {
+                            //If 'common' value is passed as a parameter, run test cases under package com.cloudant
+                            if (testsToRun.equals("common") && c.getPackage().toString().contains("common")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("mazha") && c.getPackage().toString().contains("mazha")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("datastore") && c.getPackage().toString().contains("datastore")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("indexing") && c.getPackage().toString().contains("indexing")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("query") && c.getPackage().toString().contains("query")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("replication") && c.getPackage().toString().contains("replication")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("sqlite") && c.getPackage().toString().contains("sqlite")) {
+                                classes.add(c);
+                            } else if (testsToRun.equals("util") && c.getPackage().toString().contains("util")) {
+                                classes.add(c);
+                            }
+                        } else {
+                            classes.add(c);
+                        }
```

Comment:
The above code for package filtering could be replaced with something much simpler:

```
ArrayList<Class> classes = new ArrayList<Class>();
//Check parameter for specifying which test case package to run
String testsToRun = System.getProperty("test.run-packages");
                    for (Class c : BuildConfig.classToTests) {

                        //Run all classes if JVM parameter for specifying packages does not exist
                        if (testsToRun != null) {
                        
                          if(c.getPackage().toString().contains(testsToRun)){
                              classes.add(c);
                          }

                        } else {
                            classes.add(c);
                       }
```